fix format-escapes counting count as only a number

eugenesvk opened the pull request:
while per spec it can include an identifier

count     := parameter | integer
parameter := argument '$'
argument  := integer | identifier


closes https://github.com/rust-lang/rust-enhanced/issues/546

eugenesvk committed:
fix format-escapes counting count as only a number

while per spec it can include an identifier

count     := parameter | integer
parameter := argument '$'
argument  := integer | identifier

ehuss commented on RustEnhanced.sublime-syntax:
```diff
@@ -1441,8 +1441,8 @@ contexts:
-            (\d+\$?)?             # [width]
-            (\.(\d+\$?|\*)?)?     # ['.' precision]
+               ((\d+|{{identifier}})\$?   )  ?  # [width]
+            (\.((\d+|{{identifier}})\$?|\*)?)?  # ['.' precision]
```
Thanks!

It looks like the `$` is required if it is an identifier. Can you update that?

Also, can you add some tests for this (in `syntax_test_macros.rs`).

eugenesvk replied:
updated.

Added some basic tests, but note that crrently the syntax file is too lenient and doesn't catch more granular errors within format strings like missing `$` in identifier `pad$` because it only checks the format string "in total" without splitting rules into individual items and checking that if the first item like `{` is found then it must be either empty and finish with `}` or only include valid elements like `pad$`, but fail on everything else like `pad

But at least it doesn't highlight it incorrectly, so at least there is a visual cue, though ideally there should be an "error" mark after `pad` to signal exactly the place that's missing the symbol

eugenesvk replied:
noticed the type is also too broad, was catching the missing `$` as correct,  so limited it to the few allowed chars instead of a broad ID

eugenesvk committed:
fix format-escapes having too broad of a type definition

while per spec it can include only the following:

type := '?' | 'x?' | 'X?' | 'o' | 'x' | 'X' | 'p' | 'b' | 'e' | 'E'

eugenesvk committed:
add basic tests for format-escapes

Currently the syntax file doesn't catch more granular errors within format strings like missing `$` in identifier `pad$` because it only checks the format string "in total" without splitting rules into individual items and checking that if the first item like `{` is found then it must be either empty and finish with `}` or only include valid elements like `pad$`, but fail on everything else like `pad

ehuss reviewed (approved):
Thanks!
